File: clients/ios/Lextures/Core/LMS/FileDownloadManager.swift
import Foundation

/// Authenticated download + cache helpers for course files (M3.2 / M0.2).
enum FileDownloadManager {
    static func contentURL(courseCode: String, source: CourseFileContentSource) -> URL {
        AppConfiguration.apiURL(path: CourseFileLogic.contentPath(courseCode: courseCode, source: source))
    }

    static func previewURL(courseCode: String, itemId: String) -> URL {
        AppConfiguration.apiURL(path: CourseFileLogic.previewPath(courseCode: courseCode, itemId: itemId))
    }

    static func authorizedRequest(url: URL, accessToken: String, range: String? = nil) -> URLRequest {
        var request = URLRequest(url: url)
        request.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
        request.setValue("ios", forHTTPHeaderField: "X-Platform")
        if let range {
            request.setValue("bytes=\(range)", forHTTPHeaderField: "Range")
        }
        return request
    }

    static func fetchData(
        courseCode: String,
        target: FilePreviewTarget,
        accessToken: String
    ) async throws -> Data {
        let url = contentURL(courseCode: courseCode, source: target.source)
        let request = authorizedRequest(url: url, accessToken: accessToken)
        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, (200 ... 299).contains(http.statusCode) else {
            let status = (response as? HTTPURLResponse)?.statusCode ?? 0
            throw APIError.httpStatus(status, message: nil)
        }
        return data
    }

    static func download(
        target: FilePreviewTarget,
        accessToken: String,
        offline: OfflineService
    ) async throws {
        let key = CourseFileLogic.downloadKey(courseCode: target.courseCode, target: target)
        if await offline.isDownloaded(key: key) { return }
        let data = try await fetchData(courseCode: target.courseCode, target: target, accessToken: accessToken)
        try await offline.downloadContent(
            key: key,
            data: data,
            fileName: target.displayName,
            mimeType: target.mimeType
        )
    }

    static func cachedData(
        target: FilePreviewTarget,
        offline: OfflineService
    ) async -> Data? {
        let key = CourseFileLogic.downloadKey(courseCode: target.courseCode, target: target)
        return await offline.downloadedData(key: key)
    }

    static func isDownloaded(
        target: FilePreviewTarget,
        offline: OfflineService
    ) async -> Bool {
        let key = CourseFileLogic.downloadKey(courseCode: target.courseCode, target: target)
        return await offline.isDownloaded(key: key)
    }
}

File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/FileDownloadManager.kt
package com.lextures.android.core.lms

import com.lextures.android.core.config.AppConfiguration
import com.lextures.android.core.network.ApiError
import com.lextures.android.core.offline.OfflineService
import okhttp3.OkHttpClient
import okhttp3.Request

object FileDownloadManager {
    private val http = OkHttpClient()

    fun contentUrl(courseCode: String, target: FilePreviewTarget): String =
        AppConfiguration.apiUrl(
            CourseFileLogic.contentPath(courseCode, target.source, target.sourceId),
        ).toString()

    fun previewUrl(courseCode: String, itemId: String): String =
        AppConfiguration.apiUrl(CourseFileLogic.previewPath(courseCode, itemId)).toString()

    fun authorizedRequest(url: String, accessToken: String, range: String? = null): Request {
        val builder = Request.Builder()
            .url(url)
            .header("Authorization", "Bearer $accessToken")
            .header("X-Platform", "android")
        if (range != null) {
            builder.header("Range", "bytes=$range")
        }
        return builder.build()
    }

    suspend fun fetchBytes(target: FilePreviewTarget, accessToken: String): ByteArray {
        val request = authorizedRequest(contentUrl(target.courseCode, target), accessToken)
        http.newCall(request).execute().use { response ->
            if (!response.isSuccessful) {
                throw ApiError.HttpStatus(response.code, null)
            }
            return response.body?.bytes() ?: ByteArray(0)
        }
    }

    suspend fun fetchReportCardPdf(cardId: String, accessToken: String): ByteArray {
        val url = AppConfiguration.apiUrl("/api/v1/report-cards/$cardId/pdf").toString()
        val request = authorizedRequest(url, accessToken)
        http.newCall(request).execute().use { response ->
            if (!response.isSuccessful) {
                throw ApiError.HttpStatus(response.code, null)
            }
            return response.body?.bytes() ?: ByteArray(0)
        }
    }

    suspend fun download(
        target: FilePreviewTarget,
        accessToken: String,
        offline: OfflineService,
    ) {
        val key = CourseFileLogic.downloadKey(target.courseCode, target)
        if (offline.isDownloaded(key)) return
        val data = fetchBytes(target, accessToken)
        offline.downloadContent(key, data, target.displayName, target.mimeType)
    }

    fun cachedBytes(target: FilePreviewTarget, offline: OfflineService): ByteArray? {
        val key = CourseFileLogic.downloadKey(target.courseCode, target)
        return offline.downloadedData(key)
    }

    fun isDownloaded(target: FilePreviewTarget, offline: OfflineService): Boolean {
        val key = CourseFileLogic.downloadKey(target.courseCode, target)
        return offline.isDownloaded(key)
    }
}
